source/core/slang-downstream-compiler.h

```diff
@@ -0,0 +1,403 @@
+        String value;
+    };
+
+    struct CompileOptions
+    {
+        typedef uint32_t Flags;
+        struct Flag 
+        {
+            enum Enum : Flags
+            {
+                EnableExceptionHandling = 0x01,
+                Verbose                 = 0x02,
+                EnableSecurityChecks    = 0x04,
+            };
+        };
+
+        OptimizationLevel optimizationLevel = OptimizationLevel::Default;
+        DebugInfoType debugInfoType = DebugInfoType::Standard;
+        TargetType targetType = TargetType::Executable;
+        SourceType sourceType = SourceType::CPP;
+        FloatingPointMode floatingPointMode = FloatingPointMode::Default;
+
+        Flags flags = Flag::EnableExceptionHandling;
+
+        PlatformKind platform = PlatformKind::Unknown;
+
+            /// The path/name of the output module. Should not have the extension, as that will be added for each of the target types.
+            /// If not set a module path will be internally generated internally on a command line based compiler
+        String modulePath;                  
+
+        List<Define> defines;
+
+            /// The contents of the source to compile. This can be empty is sourceFiles is set.
+            /// If the compiler is a commandLine file this source will be written to a temporary file.
+        String sourceContents;
+            /// The names/paths of source to compile. This can be empty if sourceContents is set.
+        List<String> sourceFiles;           
+
+        List<String> includePaths;
+        List<String> libraryPaths;
+    };
+
+    typedef uint32_t ProductFlags;
+    struct ProductFlag
+    {
+        enum Enum : ProductFlags
+        {
+            Debug               = 0x1,                ///< Used by debugger during execution
+            Execution           = 0x2,                ///< Required for execution
+            Compile             = 0x4,                ///< A product *required* for compilation
+            Miscellaneous       = 0x8,                ///< Anything else    
+        };
+        enum Mask : ProductFlags
+        {
+            All                 = 0xf,                ///< All the flags
+        };
+    };
+
+    enum class Product
+    {
+        DebugRun,
+        Run,
+        CompileTemporary,
+        All,
+    };
+
+        /// Get the desc of this compiler
+    const Desc& getDesc() const { return m_desc;  }
+        /// Compile using the specified options. The result is in resOut
+    virtual SlangResult compile(const CompileOptions& options, RefPtr<DownstreamCompileResult>& outResult) = 0;
+ 
+        /// Return the compiler type as name
+    static UnownedStringSlice getCompilerTypeAsText(CompilerType type);
+
+protected:
+
+    DownstreamCompiler(const Desc& desc) :
+        m_desc(desc)
+    {}
+
+    Desc m_desc;
+};
+
+class CommandLineDownstreamCompileResult : public DownstreamCompileResult
```

Comment:
Not a major design issue, but here we are combining "is invoked via command-line/`exec`" together with "produces its results via files on disk." There could in principle be API-based compilers that still write their outputs to disk files, and there could be command-line tools that can return their output without touching the file system (e.g., by writing it to `stdout`).

I think this is a good simplifying assumption, but figured I should note it in case were ever need to decouple these two aspects of command-line-ness.